tests/lib/state_test.py

```diff
@@ -138,3 +140,138 @@ def _extra_state(self):
                 {'name': 'eth1', 'state': 'up', 'type': 'unknown'}
             ]
         })
+
+
+def _get_test_ipv4_routes():
+    return [
+        {
+            Route.DESTINATION: '198.51.100.0/24',
+            Route.METRIC: 103,
+            Route.NEXT_HOP_INTERFACE: 'eth1',
+            Route.NEXT_HOP_ADDRESS: '192.0.2.1',
+            Route.TABLE_ID: 50
+        },
+        {
+            Route.DESTINATION: '203.0.113.0/24',
+            Route.METRIC: 103,
+            Route.NEXT_HOP_INTERFACE: 'eth1',
+            Route.NEXT_HOP_ADDRESS: '192.0.2.2',
+            Route.TABLE_ID: 51
+        }
+    ]
+
+
+def _get_test_ipv6_routes():
+    return [
+        {
+            Route.DESTINATION: '2001:db8:a::/64',
+            Route.METRIC: 103,
+            Route.NEXT_HOP_INTERFACE: 'eth1',
+            Route.NEXT_HOP_ADDRESS: '2001:db8:1::a',
+            Route.TABLE_ID: 50
+        },
+        {
+            Route.DESTINATION: '2001:db8:b::/64',
+            Route.METRIC: 103,
+            Route.NEXT_HOP_INTERFACE: 'eth1',
+            Route.NEXT_HOP_ADDRESS: '2001:db8:1::b',
+            Route.TABLE_ID: 51
+        }
+    ]
+
+
+class TestAssertRouteState(object):
+    parametrize_ip_ver_routes = pytest.mark.parametrize(
+        'routes',
+        [(_get_test_ipv4_routes()),
+         (_get_test_ipv6_routes())],
+        ids=['ipv4', 'ipv6'])
+
+    @parametrize_ip_ver_routes
+    def test_hash_unique(self, routes):
+        assert hash(state._Route(routes[0])) == hash(state._Route(routes[0]))
```

Comment:
You still need to check the negative:
`assert hash(state._Route(routes[0])) != hash(state._Route(routes[1]))`

Here is a buggy implementation that this test is not catching:
```
def __hash__(self):
    hash(a, b, c, d, ...)
```

Reply:
Your understanding of `hash()` is wrong. The hash function can generate the same result for two
different data. So we can never know for sure we can get different hash results for different tuple.

Reply:
Please respond on how you cover the bug example.
Although in theory the negative test may fail, it is unlikely. You can add a comment that warns for it if you like.